server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java

```diff
@@ -126,6 +145,67 @@ public void handle(Request request, Response response) throws Exception {
     response.noContent();
   }
 
+  private void doHandle(DbSession dbSession, SetRequest request) {
+    Optional<ComponentDto> component = searchComponent(dbSession, request);
+    checkPermissions(component);
+
+    PropertyDefinition definition = propertyDefinitions.get(request.getKey());
+
+    if (!request.getFieldsValues().isEmpty()) {
+      doHandlePropertySet(dbSession, request, definition, component);
+    } else {
+      validate(request, component);
+      dbClient.propertiesDao().insertProperty(dbSession, toProperty(request, component));
+      if (!component.isPresent()) {
+        settingsChangeNotifier.onGlobalPropertyChange(request.getKey(), persistedValue(request));
+      }
+    }
+  }
+
+  private void doHandlePropertySet(DbSession dbSession, SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
+    validatePropertySet(request, definition, component);
+
+    Set<String> fieldKeys = definition.fields().stream()
+      .map(PropertyFieldDefinition::key)
+      .collect(Collectors.toCollection(LinkedHashSet::new));
+    String inlinedFieldKeys = COMMA_JOINER.join(fieldKeys);
+    String key = persistedKey(request);
+    Long componentId = component.isPresent() ? component.get().getId() : null;
+
+    dbClient.propertiesDao().insertProperty(dbSession, new PropertyDto().setKey(key).setValue(inlinedFieldKeys).setResourceId(componentId));
+
+    List<Map<String, String>> fieldsValues = request.getFieldsValues();
+    IntStream.rangeClosed(1, fieldsValues.size())
+      .boxed()
+      .flatMap(i -> fieldsValues.get(i - 1).entrySet().stream().map(entry -> new KeyValue(key + "." + i + "." + entry.getKey(), entry.getValue())))
+      .forEach(keyValue -> dbClient.propertiesDao().insertProperty(dbSession, new PropertyDto().setKey(keyValue.key).setValue(keyValue.value).setResourceId(componentId)));
+  }
+
+  private void validatePropertySet(SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
+    checkRequest(definition != null, "Setting '%s' is undefined", request.getKey());
+    checkRequest(PropertyType.PROPERTY_SET.equals(definition.type()), "Parameter '%s' is used for setting of property set type only", PARAM_FIELDS_VALUES);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) NullPointerException might be thrown as 'definition' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)